NorthstarDLL/mods/autodownload/moddownloader.cpp

```diff
@@ -0,0 +1,523 @@
+#include "moddownloader.h"
+#include <rapidjson/fwd.h>
+#include <mz_strm_mem.h>
+#include <mz.h>
+#include <mz_strm.h>
+#include <mz_zip.h>
+#include <mz_compat.h>
+#include <thread>
+#include <future>
+#include <bcrypt.h>
+#include <winternl.h>
+#include <fstream>
+
+ModDownloader* g_pModDownloader;
+
+ModDownloader::ModDownloader()
+{
+	spdlog::info("Mod downloader initialized");
+}
+
+size_t WriteToString(void* ptr, size_t size, size_t count, void* stream)
+{
+	((std::string*)stream)->append((char*)ptr, 0, size * count);
+	return size * count;
+}
+
+void ModDownloader::FetchModsListFromAPI()
+{
+	std::thread requestThread(
+		[this]()
+		{
+			CURLcode result;
+			CURL* easyhandle;
+			rapidjson::Document verifiedModsJson;
+			std::string url = MODS_LIST_URL;
+
+			curl_global_init(CURL_GLOBAL_ALL);
+			easyhandle = curl_easy_init();
+			std::string readBuffer;
+
+			// Fetching mods list from GitHub repository
+			curl_easy_setopt(easyhandle, CURLOPT_CUSTOMREQUEST, "GET");
+			curl_easy_setopt(easyhandle, CURLOPT_TIMEOUT, 30L);
+			curl_easy_setopt(easyhandle, CURLOPT_URL, url.c_str());
+			curl_easy_setopt(easyhandle, CURLOPT_FAILONERROR, 1L);
+			curl_easy_setopt(easyhandle, CURLOPT_WRITEDATA, &readBuffer);
+			curl_easy_setopt(easyhandle, CURLOPT_WRITEFUNCTION, WriteToString);
+			result = curl_easy_perform(easyhandle);
+
+			if (result == CURLcode::CURLE_OK)
+			{
+				spdlog::info("Mods list successfully fetched.");
+			}
+			else
+			{
+				spdlog::error("Fetching mods list failed: {}", curl_easy_strerror(result));
+				goto REQUEST_END_CLEANUP;
+			}
+
+			// Load mods list into local state
+			spdlog::info("Loading mods configuration...");
+			verifiedModsJson.Parse(readBuffer);
+			for (auto i = verifiedModsJson.MemberBegin(); i != verifiedModsJson.MemberEnd(); ++i)
+			{
+				std::string name = i->name.GetString();
+				std::string dependency = i->value["DependencyPrefix"].GetString();
+
+				std::unordered_map<std::string, VerifiedModVersion> modVersions;
+				rapidjson::Value& versions = i->value["Versions"];
```

Comment:
```suggestion
			for (auto& i : verifiedModsJson.GetObject())
			{
				std::string name = i.name.GetString();
				std::string dependency = i.value["DependencyPrefix"].GetString();

				std::unordered_map<std::string, VerifiedModVersion> modVersions;
				rapidjson::Value& versions = i.value["Versions"];
```

Reply:
Visual studio apparently doesn't like it:
`class "rapidjson::GenericDocument<...>" has no member "GetObjectW"`

Reply:
`GetObjectW`? Where is that coming from?

Reply:
`#define GetObject GetObjectW`

Reply:
> `#define GetObject GetObjectW`

Where is this?

Reply:
From rapidjson I guess?

Reply:
Maybe not

![image](https://github.com/R2Northstar/NorthstarLauncher/assets/11993538/56261084-5f67-48d5-9377-da8416882a4c)

Reply:
`scriptjson.cpp` i think solves this by doing this at the top of the file:

```
#ifdef _MSC_VER
#undef GetObject // fuck microsoft developers
#endif
```
so perhaps that would work here as well

Reply:
Ok so, this is just not worth doing at this point. whatever